docs/proposals/ingress-l4.md

```diff
@@ -0,0 +1,346 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Ingress L4 Proposal
+
+## Abstract
+
+A proposal to add support for TCP and UDP to the Ingress Resource.
+
+Existing issue regarding Ingress and TCP
+* Ingress and TCP [#23291](https://github.com/kubernetes/kubernetes/issues/23291)
+
+
+## Use Cases
+
+1. Be able to accept and route traffic for arbitrary TCP ports
+1. Be able to accept and route traffic for arbitrary UDP ports
+
+
+## Motivation
+
+When deploying and exposing non-http(s) based services, Kubernetes users are
+limited to NodePort and service type=LoadBalancer as methods for exposing these
+services using interfaces supported by the Kubernetes API.
+
+Other methods exist, via contributed code, for exposing non-http(s) based services:
+
+* [keepalived-vip](https://github.com/kubernetes/contrib/tree/master/keepalived-vip)
+* NGINX Ingress Controller
+  [nginx-ingress](https://github.com/kubernetes/contrib/tree/master/ingress/controllers/nginx)
+
+However these implementations rely on a ConfigMap being populated with a
+mapping of destination endpoint to service.  A single common interface for
+describing all Ingress would is highly desirable.
+
+
+## Design Considerations
+
+### State of L4 Protocol Support in Cloud Providers
+
+L4 support is not uniform across all Cloud Providers, specifically TLS
+termination is lacking in several Cloud Providers.
+
+|           | TCP | UDP | TLS |
+| --------- | --- | --- | --- |
+| AWS       |  x  |     |  x  |
+| GCE       |  x  |  x  | (1) |
```

Comment:
Correct, TCP and UDP are GA but TLS is alpha according to https://cloud.google.com/compute/docs/load-balancing/tcp-ssl/

I'm aware HTTPS is supported and GA, but since this proposal is for L4, I wanted to leave HTTP/HTTPS information out so as not to confuse the discussion, but I can add that and add a description of the table if that would help clarify